megamek/src/megamek/common/PilotSPAHelper.java

```diff
@@ -0,0 +1,95 @@
+package megamek.common;
+
+import megamek.common.weapons.autocannons.LBXACWeapon;
+import megamek.common.weapons.autocannons.UACWeapon;
+import megamek.common.weapons.bayweapons.BayWeapon;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * This class contains helper methods for Special Pilot Abilities.
+ */
+public final class PilotSPAHelper {
+
+    /** @return True when the given Mounted equipment is a valid choice for the Weapons Specialist SPA. */
+    public static boolean isWeaponSpecialistValid(Mounted mounted) {
+        return isWeaponSpecialistValid(mounted.getType());
+    }
+
+    /** @return True when the given EquipmentType is a valid choice for the Weapons Specialist SPA. */
+    public static boolean isWeaponSpecialistValid(EquipmentType equipmentType) {
+        return (equipmentType instanceof WeaponType) && !(equipmentType instanceof BayWeapon)
+                && !equipmentType.hasFlag(WeaponType.F_AMS) && !equipmentType.is("Screen Launcher");
+    }
+
+    /**
+     * Returns a List of distinct (each occuring only once) weapon names of weapons present on the given
+     * Entity that are valid choices for the Weapon Specialist SPA.
+     *
+     * @return A list of weapon names from the given Entity that are valid choices for the Weapon Specialist SPA
+     */
+    public static List<String> weaponSpecialistValidWeaponNames(Entity entity) {
+        return entity.getTotalWeaponList().stream()
+                .map(Mounted::getType)
+                .filter(PilotSPAHelper::isWeaponSpecialistValid)
+                .map(EquipmentType::getName)
+                .distinct()
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Returns a List of weapons from those present on the given Entity that are valid choices for the
+     * Weapon Specialist SPA. Unlike {@link #weaponSpecialistValidWeaponNames(Entity)}, weapons
+     * appear in this list as often as they are present on the given Entity.
+     *
+     * @return A list of weapons from the given Entity that are valid choices for the Weapon Specialist SPA
+     */
+    public static List<Mounted> weaponSpecialistValidWeapons(Entity entity) {
+        return entity.getTotalWeaponList().stream()
+                .filter(PilotSPAHelper::isWeaponSpecialistValid)
+                .collect(Collectors.toList());
+    }
+
+    /** @return True when the given Mounted equipment is a valid choice for the Sandblaster SPA. */
+    public static boolean isSandblasterValid(Mounted mounted) {
+        return isSandblasterValid(mounted.getType());
+    }
+
+    /** @return True when the given EquipmentType is a valid choice for the Sandblaster SPA. */
+    public static boolean isSandblasterValid(EquipmentType equipmentType) {
+        return (equipmentType instanceof WeaponType)
+                && ((equipmentType instanceof UACWeapon) || (equipmentType instanceof LBXACWeapon)
```

Comment:
One comment here, which is that the "rapid fire autocannons" weapon allows standard ACs to behave like UACs (at the risk of exploding), so I think standard autocannons should be included in this.

Reply:
As a suggestion, I have added a GameOptions parameter here based on which all ACs may be allowed. Both MM and MHQ can supply GameOptions and so it'll reflect the state of the game/campaign. The GameOptions can be nulled if necessary but I have not added versions without the additional parameter as we'd really want the GameOpts to be given. I think MML won't need these methods.

Aside from that I've removed Streaks from Sandblaster as they don't use the cluster table and C3 from the weapon specialist (hehe).

Reply:
@NickAragua Any thoughts?

Reply:
Sorry, lost track of this one.

I think there are some situations (mostly involving high powered ECM as I recall) where Streak launchers revert to cluster rolls, so there may be a case for Sandblaster applying to them.